File: src/Keryhe.Telemetry.Client/Services/SearchQueryParser.cs
namespace Keryhe.Telemetry.Client.Services;

public record SearchTerm(
    bool IsAttributeFilter,
    string? Key,
    string? Value,
    bool IsExactMatch,
    string? FreeText);

public record ParsedSearchQuery(
    bool IsTraceIdSearch,
    string? TraceId,
    List<SearchTerm> Terms);

public static class SearchQueryParser
{
    public static ParsedSearchQuery Parse(string searchText)
    {
        if (string.IsNullOrWhiteSpace(searchText))
            return new ParsedSearchQuery(false, null, new List<SearchTerm>());

        var trimmed = searchText.Trim();

        // Trace ID: 32 hex chars, no spaces, =, :, or quotes
        if (trimmed.Length == 32 &&
            !trimmed.Contains(' ') &&
            !trimmed.Contains('=') &&
            !trimmed.Contains(':') &&
            !trimmed.Contains('"') &&
            IsHexString(trimmed))
        {
            return new ParsedSearchQuery(true, trimmed, new List<SearchTerm>());
        }

        var rawTerms = trimmed.Split(new[] { " AND " }, StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        var terms = new List<SearchTerm>();

        foreach (var raw in rawTerms)
        {
            var term = raw.Trim();

            if (term.Contains('='))
            {
                var parts = term.Split('=', 2);
                terms.Add(new SearchTerm(true, parts[0].Trim(), parts[1].Trim().Trim('"', '\''), true, null));
            }
            else if (term.Contains(':'))
            {
                var parts = term.Split(':', 2);
                terms.Add(new SearchTerm(true, parts[0].Trim(), parts[1].Trim().Trim('"', '\''), false, null));
            }
            else
            {
                terms.Add(new SearchTerm(false, null, null, false, term.Trim('"')));
            }
        }

        return new ParsedSearchQuery(false, null, terms);
    }

    private static bool IsHexString(string s)
    {
        foreach (var c in s)
        {
            if (!Uri.IsHexDigit(c))
                return false;
        }
        return true;
    }
}

File: src/telemetry-client/src/app/shared/utils/search-query.parser.ts
// Client-side port of Blazor's SearchQueryParser.cs — shared by the Logs and Traces pages.
// Supports: free text, exact `key=value`, contains `key:value`, `AND` between terms,
// a leading `-` to negate/exclude a term, and trace-id detection (32 hex chars =>
// fetch by trace, date range ignored).

export interface SearchTerm {
  isAttributeFilter: boolean;
  key?: string;
  value?: string;
  isExactMatch: boolean;
  freeText?: string;
  /** True when the term is prefixed with `-` (exclude matches rather than include). */
  negate?: boolean;
}

/** Build the `key=value` / `key:value` (optionally `-`-prefixed) text for a facet click. */
export function buildAttributeTerm(key: string, value: string, exclude = false): string {
  // Contains-match (`:`) keeps facet clicks consistent with the search box default.
  return `${exclude ? '-' : ''}${key}:${quoteIfNeeded(value)}`;
}

function quoteIfNeeded(value: string): string {
  return /[\s:=]/.test(value) ? `"${value}"` : value;
}

export interface ParsedSearchQuery {
  isTraceIdSearch: boolean;
  traceId?: string;
  terms: SearchTerm[];
}

const HEX_RE = /^[0-9a-fA-F]+$/;

function trimQuotes(s: string): string {
  return s.replace(/^['"]+|['"]+$/g, '');
}

export function parseSearchQuery(searchText: string): ParsedSearchQuery {
  if (!searchText || !searchText.trim()) {
    return { isTraceIdSearch: false, terms: [] };
  }

  const trimmed = searchText.trim();

  // Trace ID: exactly 32 hex chars, no spaces, =, :, or quotes.
  if (
    trimmed.length === 32 &&
    !trimmed.includes(' ') &&
    !trimmed.includes('=') &&
    !trimmed.includes(':') &&
    !trimmed.includes('"') &&
    HEX_RE.test(trimmed)
  ) {
    return { isTraceIdSearch: true, traceId: trimmed, terms: [] };
  }

  const rawTerms = trimmed
    .split(' AND ')
    .map((t) => t.trim())
    .filter((t) => t.length > 0);

  const terms: SearchTerm[] = rawTerms.map((raw) => {
    // A leading `-` (before a key/value or free text) excludes matches.
    const negate = raw.startsWith('-');
    const term = negate ? raw.slice(1).trim() : raw;
    const eq = term.indexOf('=');
    const colon = term.indexOf(':');

    if (eq >= 0) {
      return {
        isAttributeFilter: true,
        key: term.slice(0, eq).trim(),
        value: trimQuotes(term.slice(eq + 1).trim()),
        isExactMatch: true,
        negate,
      };
    }
    if (colon >= 0) {
      return {
        isAttributeFilter: true,
        key: term.slice(0, colon).trim(),
        value: trimQuotes(term.slice(colon + 1).trim()),
        isExactMatch: false,
        negate,
      };
    }
    return {
      isAttributeFilter: false,
      isExactMatch: false,
      freeText: trimQuotes(term),
      negate,
    };
  });

  return { isTraceIdSearch: false, terms };
}
